[rpc] poll for timed out futures in process group agent

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29601 [rpc] poll for timed out futures in process group agent**

Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

The existing `futureTimeouts_` map is reworked so that the key is now the time (in ms) that the RPC would timeout, instead of when the RPC began. The reason for this is that it allows us to order futures which have different timeouts, which is a feature that we'd want to support in the future. A helper function `getRPCEndTime` can be used to retrieve the key into this map. 

Since we'd like to be able to quickly stop checking for timed out RPCs after looking at the first entry in the `futureTimeouts_` map, we cannot key by beginning time (an RPC with a much smaller timeout may later be created) or by the timeout itself (since an RPC with a large timeout would not be at the beginning, even if it is about to time out).

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

rohan-varma committed:
[wip][rpc] poll for timed out futures in process group agent

Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs at a pre-set interval, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs at a pre-set interval, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs at a pre-set interval, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs at a pre-set interval, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

pietern commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -130,6 +130,7 @@ ProcessGroupAgent::ProcessGroupAgent(
       nextId_(0),
       sendMutexes_(pg_->getSize()),
       threadPool_(numSendRecvThreads) {
+  shutdown_.store(false);
```
Add to the member initializer list.

pietern commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +458,53 @@ void ProcessGroupAgent::listenLoop() {
 
+void ProcessGroupAgent::pollTimedOutRPCs() {
+  while (!shutdown_.load()) {
+    std::this_thread::sleep_for(std::chrono::seconds(5));
```
You can determine the time to wait by looking at the first entry in the ordered futures map. Then you either wait that long, or until you're woken up by a condition variable if there are no futures to begin with.

rohan-varma replied:
@pietern, for this suggestion, should we add a new condition variable that wakes up this thread whenever there is a new future created? There is currently a `futureCV` but this notifies when a future has been received and marked as completed.

pietern replied:
Yes, that sounds good. In order to guarantee correctness we have to do this. There are all sorts of edge cases that we'd fail to cover otherwise. We only need to wait on this one if there are no futures to begin with, though. If there are, we can use the timeout of the first one. This assumes that it never happens that you first call fut1 with timeout=5 and right after that call fut2 with timeout=1 -- it assumes the absolute timestamp of timeouts will only increase. If we were to allow for this case, the logic becomes more complex and contentious, because you'd have to wake up every time a future is created, which seems wasteful to me.

pietern commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +458,53 @@ void ProcessGroupAgent::listenLoop() {
+        auto diffTime =
+            std::chrono::duration_cast<std::chrono::milliseconds>(
+                std::chrono::steady_clock::now().time_since_epoch()) -
+            startTime;
```
All these locals are const.

rohan-varma committed:
Update on "[wip][rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs at a pre-set interval, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs at a pre-set interval, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs at a pre-set interval, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs at a pre-set interval, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs at a pre-set interval, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][rpc] poll for timed out futures in process group agent"


(not yet ready for a review) Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs at a pre-set interval, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][rpc] poll for timed out futures in process group agent"


(not yet ready for a review) Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs at a pre-set interval, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][rpc] poll for timed out futures in process group agent"


(not yet ready for a review) Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs at a pre-set interval, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][rpc] poll for timed out futures in process group agent"


(not yet ready for a review) Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs at a pre-set interval, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][rpc] poll for timed out futures in process group agent"


(not yet ready for a review) Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs at a pre-set interval, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/future_message.h:
```diff
@@ -23,6 +23,15 @@ struct TORCH_API FutureMessage final {
   void addCallback(const Callback& callback);
 
+  // Get the destination rank of this future.
+  inline int dst() const {
```
cc @zhaojuanmao @mrshenli I had to add the dest rank to the future object so that `hasPendingMessage()` in pg agent can return correctly (it checks to ensure that all send/recv counts are consistent). I'm not sure if we want to add this to the future, is there a better way to do this?

rohan-varma replied:
This is now removed.

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +476,101 @@ void ProcessGroupAgent::listenLoop() {
+    {
+      std::unique_lock<std::mutex> lock(futureTimeoutMutex_);
+      if (futureTimeoutCV_.wait_for(
+              lock, sleepTime, [this] { return shutdown_.load(); })) {
```
@pietern, let me know if this wasn't exactly what you were suggesting. Also, I think we'd need the sleep to be interruptible, since we wouldn't want to sleep if process group agent is shutting down.

jjlilley replied:
You might not want the predicate looping here, but rather just wait_for (or wait_until)
The reason is because if there's a newly added entry to the table with a sooner expiration time, you want to wake up and recalculate

pietern replied:
I expected to collapse the top block and this one into a single one. The timeouts are stored in the future timeout map, so why do we need access to the futures themselves to figure out how long we should sleep?

rohan-varma replied:
@jjlilley Are we concerned that a lot of futures will be created (with say, the same timeout) and thus this thread would be woken up too frequently? This was the reasoning why I had that predicate.

jjlilley replied:
@rohan-varma 
The thread gets woken up the same number of times with/without the predicate...

The predicate ends up just being a while loop syntactic sugar.

In the actual generated code, it is [C pseudo-code] something like 

   if (!pred()) {
     auto absolute_timeout = now() + timeout;
     do {
        if (pthread_cond_timedwait(&cv, &lock->mutex, absolute_timeout) == -1 && errno == ETIMEOUT)  break;
      } while (!pred());
  }

Hence, adding a predicate doesn't really "save" anything, and you might as well check if you've hit an expiration after the cv wakes up the thread. And in this particular case, having the pred was preventing it from waking up from an infinite timeout, if you wanted one for 0 futures.

If you have a grouping of futures with effectively the exact same absolute timeout, your logic should effectively be able to expire the all at once anyway, so it's not like there's a single wakeup per future. You really only need to notify the condvar if you
  - go from 0 futures to 1
  - insert an entry with a timeout lower than your current earliest timeout
  - shut down

rohan-varma replied:
@pietern The two different blocks are now collapsed in the latest update.

rohan-varma replied:
@jjlilley right now we don't support per-RPC timeouts, so the 2nd condition can't happen. I changed it so that we only notify in the 1st and 3rd conditions you mentioned.

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +476,101 @@ void ProcessGroupAgent::listenLoop() {
+      // which will result in busy polling for the lock If rpcTimeout_ is 0,
+      // should we have a default sleep interval here?
+      sleepTime =
+          std::max(rpcTimeout_ - runningTime, std::chrono::milliseconds(0));
```
Right now, if the `futureTimeouts_` map is empty, or `rpcTimeout_` is set to 0, we basically won't sleep and simply go back to the beginning of the loop. This could result in a lot of lock contention, reducing efficiency. Should we have some default sleep time for this thread due to this?

rohan-varma replied:
Thanks for the suggestion, this makes the code a lot simpler.

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +475,101 @@ void ProcessGroupAgent::listenLoop() {
+      // Estimate amount of time the first future will time out in, and sleep
+      // for that long.
+      if (futureTimeouts_.empty()) {
+        continue;
```
Don't busy spin here, just use some equivalent of infinity (e.g. a year is close enough), and wait in the condvar.

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +475,101 @@ void ProcessGroupAgent::listenLoop() {
+  // shutting down.
+  {
+    std::unique_lock<std::mutex> lock(futureTimeoutMutex_);
+    futureTimeoutCV_.wait(lock);
```
This block has been removed and consolidated into the inner block.

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +475,101 @@ void ProcessGroupAgent::listenLoop() {
+      // Estimate amount of time the first future will time out in, and sleep
+      // for that long.
+      if (futureTimeouts_.empty()) {
+        continue;
```
This should now be fixed with @jjlilley's suggestions.

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -281,10 +292,15 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
       futureTimeouts_[futureStartTime].push_back(requestId);
+      // Signal watchdog for future timeouts to begin
+      if (futures_.size() == 1) {
+        futureTimeoutCV_.notify_one();
```
We will also want to notify if the new timeout is earlier than the current earliest.

I suppose that's not possible yet because all timeouts are probably the same... but might be worth either 
  - adding a quick comment about the current assumption that all the timeouts are the same, and to add a check there if that changes
  - or just adding the check anyway (maybe just a couple lines)

rohan-varma replied:
If we wanted to implement this somewhere here, it should be something like 

```
if (futures_.size() == 1 ||
          getRPCRemainingTime(futureStartTime, futureInfo.timeout_) <
              getRPCRemainingTime(
                  futureTimeouts_.begin()->first,
                  futures_[futureTimeouts_.begin()->second.front()].timeout_)) {
                   notify()
}
```

I'm leaving just a note for now because it requires a good amount of reworking. For example, we'd also need to change the way we key into the `futureTimeouts_` map, since the key for that is the absolute start time, and we'd want to change it to something like a relative time that takes into account the specific `rpcTimeout` the RPC has.

pietern replied:
@jjlilley You're right for correctness, but since we have a single global RPC timeout today, they will be monotonically increasing. If we do add per-RPC timeouts (which needs to be thoroughly weighed), then we should put in additional work to avoid waking up this thread every time we add a future.

jjlilley replied:
Except if setRpcTimeout() is called after starting (e.g. cut from 1 min to 5 seconds), which is technically possible via the python bindings, though maybe not recommended/supported...

rohan-varma replied:
Good point, thanks for calling this out. I'm going to make this method private for now until we can have a discussion about how/if we want to support this.

jjlilley replied:
Thanks for humoring me :)
(I've been bit by this sort of issue in the past)

jjlilley reviewed (commented):
Nice updates. A couple quick comments..

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +474,90 @@ void ProcessGroupAgent::listenLoop() {
+    // could invoke callbacks.
+    if (!timedOutFutures.empty()) {
+      std::stringstream ss;
+      ss << "Future ran for more than " << rpcTimeout_.count()
```
nit: "Rpc ran for for" instead of Future

pritamdamania87 commented on test/rpc_test.py:
```diff
@@ -1044,6 +1044,26 @@ def test_set_rpc_timeout(self):
+            fut.wait()
+
+        # future should run to completion if the timeout is longer.
+        dist.barrier()
```
Why do we need these barriers? Also, don't we need to initialize the process group if we need these barriers? I don't think `dist_init` initializes the default process group for all backends.

rohan-varma replied:
I thought there might be issues with different timeouts existing across different workers. Though if this happens, the test will still work, because the timeout on the local worker is always used

pritamdamania87 replied:
I see, but we still need to initialize the process group right? This test probably won't work for non process group backends?

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -84,6 +84,9 @@ std::vector<int64_t> ProcessGroupAgent::MessageCounter::snapshot() {
 ////////////////////////  ProcessGroupAgent  /////////////////////////////////
 
+constexpr std::chrono::milliseconds INFINITY_SLEEP_INTERVAL =
+    std::chrono::hours(8760);
```
How do we get 8760? Can we do this instead `std::chrono::duration<int64_t, std::nano>(std::numeric_limits<int64_t>::max())`

rohan-varma replied:
I don't think this works with `std::chrono::milliseconds` (at least this particular line). How about `const INTERVAL = std::chrono::seconds(std::numeric_limits<int64_t>::max()` or something of that nature?

pritamdamania87 replied:
I think you will need to use std::chrono::nanoseconds since that is the only one that supports 64 bit value (I don't think int64_t max would work with seconds). See `Helper Types` here: https://en.cppreference.com/w/cpp/chrono/duration.

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -392,24 +408,33 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
-                std::find(futuresAtTime.begin(), futuresAtTime.end(), id));
-            if (futuresAtTime.size() == 0) {
+            auto it = std::find(futuresAtTime.begin(), futuresAtTime.end(), id);
+            TORCH_CHECK(
```
nit: `TORCH_INTERNAL_ASSERT`

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +474,90 @@ void ProcessGroupAgent::listenLoop() {
+      }
+      futureTimeoutCV_.wait_for(lock, sleepTime);
+    }
+
```
Can we add a check here that if `rpcTimeout_.count() == 0` we just continue and go back to sleeping? We don't need to call `processTimedOutFutures` right?

jjlilley replied:
But if rpcTimeout_.count() == 0, we're basically rarely/never waking up, or hitting this codepath, right?
(maybe I'm missing something?)

rohan-varma replied:
This should be the case, since we would run once, sleep for a long time, and wake up when shutting down.

pritamdamania87 replied:
Wouldn't the `notify_one` in the send function wake us up every time there is on outstanding RPC?

jjlilley replied:
No, notify_one in send() only needs to wake this up when there's a sooner timeout that the present one.
(note that the if size==1 codepath, i.e. it was previously empty)

rohan-varma replied:
We only call notify on `futureTimeoutCV` when there is the first future or when shutting down.

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +474,90 @@ void ProcessGroupAgent::listenLoop() {
+    for (auto it = futureTimeouts_.begin(); it != futureTimeouts_.end();
+         /* intentional no increment */) {
+      const std::chrono::milliseconds startTime = it->first;
+      const std::vector<int64_t> futureIDs = it->second;
```
nit: const reference?

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +474,90 @@ void ProcessGroupAgent::listenLoop() {
+      } else {
+        for (const auto& futureID : futureIDs) {
+          auto it = futures_.find(futureID);
+          TORCH_CHECK(
```
nit: TORCH_INTERNAL_ASSERT, TORCH_CHECK is mostly to indicate that the user made some error.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -122,7 +127,7 @@ class TORCH_API RpcAgent {
   const std::string workerName_;
   const std::unique_ptr<RequestCallback> cb_;
-  const std::chrono::milliseconds rpcTimeout_;
+  std::chrono::milliseconds rpcTimeout_;
```
Doesn't this need to be atomic since `setRpcTimeout` could be called from a different thread while the watchdog thread is reading from it?

rohan-varma replied:
@pritamdamania87 The set function would only be invoked via pybind from a python call, and the GIL is not dropped, so should we be fine for thread safety? I don't anticipate needing to call `setRPCTimeout` from C++.

jjlilley replied:
it might just be worth grabbing the lock in setRpcTimeout() ... compared with all the other things going on, this should be super cheap and help us avoid some unexpected corner-cases.

pritamdamania87 replied:
@rohan-varma We probably hold the GIL for writing (setRpcTimeout), but no lock is held while reading.

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -392,24 +413,33 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
+              return;
+            }
+
+            const auto& futureInfo = futures_[id];
```
we can actually cache the above iterator lookup and use it instead of looking up again

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +479,92 @@ void ProcessGroupAgent::listenLoop() {
+    }
+    futureTimeoutCV_.wait_for(lock, sleepTime);
+
+    if (rpcTimeout_.load().count() == 0) {
```
The checks for `rpcTimeout` and `shutdown` doesn't need the lock, but it is needed in `processTimedOutFutures`. Should we do a `lock.unlock()` here and re-acquire it before calling `processTimedOutFutures`? Not sure if the performance difference would be meaningful

pritamdamania87 replied:
Don't worry about it since those checks are pretty cheap.

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

pritamdamania87 commented on test/rpc_test.py:
```diff
@@ -1083,20 +1083,21 @@ def test_set_rpc_timeout(self):
         rpc.join_rpc()
 
     @dist_init
+    @requires_process_group_agent("PROCESS_GROUP rpc backend specific test, skip")
```
Can we cut a task to enable this for other backends?

rohan-varma replied:
Done

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -84,8 +84,8 @@ std::vector<int64_t> ProcessGroupAgent::MessageCounter::snapshot() {
-constexpr std::chrono::milliseconds INFINITY_SLEEP_INTERVAL =
-    std::chrono::hours(8760);
+const std::chrono::milliseconds INFINITY_SLEEP_INTERVAL =
+    std::chrono::hours(std::numeric_limits<int64_t>::max());
```
Looks like there is https://en.cppreference.com/w/cpp/chrono/duration/max that we can use directly.

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -477,25 +482,23 @@ void ProcessGroupAgent::listenLoop() {
+      sleepTime = INFINITY_SLEEP_INTERVAL;
+    } else {
+      const auto remainingTime = getRPCRemainingTime(
+          futureTimeouts_.begin()->first, rpcTimeout_.load());
```
Shouldn't we use the rpcTimeout we store in `FutureInfo`?

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -524,40 +528,43 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
+    const std::chrono::milliseconds& startTime = it->first;
+    const std::vector<int64_t>& futureIDs = it->second;
+    const auto remainingTime =
+        getRPCRemainingTime(startTime, rpcTimeout_.load());
```
Use the timeout stored in `FutureInfo`?

pritamdamania87 reviewed (changes requested):
Looks good overall, just a few comments that I feel we need to address.

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -97,8 +97,14 @@ class TORCH_API RpcAgent {
   // Retrieve the timeout for all RPCs.
-  inline const std::chrono::milliseconds& getRpcTimeout() const {
-    return rpcTimeout_;
+  inline const std::chrono::milliseconds getRpcTimeout() const {
```
remove the const from the return type?

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -97,8 +97,14 @@ class TORCH_API RpcAgent {
+  // Set the timeout for all RPCs
+  inline void setRpcTimeout(const std::chrono::milliseconds& rpcTimeout) {
+    rpcTimeout_ = rpcTimeout;
+    rpcTimeout_.store(rpcTimeout);
```
you don't need two store calls here :)

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

The existing `futureTimeouts_` map is reworked so that the key is now the time (in ms) that the RPC would timeout, instead of when the RPC began. The reason for this is that it allows us to order futures which have different timeouts, which is a feature that we'd want to support in the future. A helper function `getRPCEndTime` can be used to retrieve the key into this map.

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma commented:
I reworked the PR a bit to ensure that if a new future with a smaller TTL is created, it is appropriately inserted into the beginning of the map, and also implemented @jjlilley's suggestion to notify on the `futureTimeoutCV` if such an event occurs. The main change is that the `futureTimeouts_` map is now keyed by the expiry time of the RPC, instead of the beginning time (see PR for explanation).

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +484,99 @@ void ProcessGroupAgent::listenLoop() {
+      futureTimeoutCV_.wait_until(
+          lock, std::chrono::time_point<std::chrono::system_clock>::max());
+    } else {
+      futureTimeoutCV_.wait_for(lock, sleepTime);
```
I encountered some overflow errors when trying to pass in `INFINITE_TIMEOUT` into `wait_for`. Not sure if there's a better way to get around this.

jjlilley replied:
In this case, I think it's idiomatic to just use wait():

  if (sleepTime == INFINITE_TIMEOUT) {
       futureTimeoutCV_.wait();
  } else {
     futureTimeoutCV_.wait_for(lock, sleepTime);
  }

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +484,99 @@ void ProcessGroupAgent::listenLoop() {
+  return timedOutFutures;
+}
+
+const std::chrono::milliseconds ProcessGroupAgent::getRPCRemainingTime(
```
I feel like these functions don't really belong in PG agent since they are rpc/future specific. As a follow up we should figure out where these sort of functions should go

jjlilley replied:
right, this could even be in private namespace {}, but this is fine too for now

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

The existing `futureTimeouts_` map is reworked so that the key is now the time (in ms) that the RPC would timeout, instead of when the RPC began. The reason for this is that it allows us to order futures which have different timeouts, which is a feature that we'd want to support in the future. A helper function `getRPCEndTime` can be used to retrieve the key into this map. 

Since we'd like to be able to quickly stop checking for timed out RPCs after looking at the first entry in the `futureTimeouts_` map, we cannot key by beginning time (an RPC with a much smaller timeout may later be created) or by the timeout itself (since an RPC with a large timeout would not be at the beginning, even if it is about to time out).

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

The existing `futureTimeouts_` map is reworked so that the key is now the time (in ms) that the RPC would timeout, instead of when the RPC began. The reason for this is that it allows us to order futures which have different timeouts, which is a feature that we'd want to support in the future. A helper function `getRPCEndTime` can be used to retrieve the key into this map. 

Since we'd like to be able to quickly stop checking for timed out RPCs after looking at the first entry in the `futureTimeouts_` map, we cannot key by beginning time (an RPC with a much smaller timeout may later be created) or by the timeout itself (since an RPC with a large timeout would not be at the beginning, even if it is about to time out).

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

The existing `futureTimeouts_` map is reworked so that the key is now the time (in ms) that the RPC would timeout, instead of when the RPC began. The reason for this is that it allows us to order futures which have different timeouts, which is a feature that we'd want to support in the future. A helper function `getRPCEndTime` can be used to retrieve the key into this map. 

Since we'd like to be able to quickly stop checking for timed out RPCs after looking at the first entry in the `futureTimeouts_` map, we cannot key by beginning time (an RPC with a much smaller timeout may later be created) or by the timeout itself (since an RPC with a large timeout would not be at the beginning, even if it is about to time out).

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

pritamdamania87 reviewed (approved):
Looks good to me, @jjlilley could you take another look as well? Thanks!

jjlilley reviewed (approved):
LGTM as well

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

The existing `futureTimeouts_` map is reworked so that the key is now the time (in ms) that the RPC would timeout, instead of when the RPC began. The reason for this is that it allows us to order futures which have different timeouts, which is a feature that we'd want to support in the future. A helper function `getRPCEndTime` can be used to retrieve the key into this map. 

Since we'd like to be able to quickly stop checking for timed out RPCs after looking at the first entry in the `futureTimeouts_` map, we cannot key by beginning time (an RPC with a much smaller timeout may later be created) or by the timeout itself (since an RPC with a large timeout would not be at the beginning, even if it is about to time out).

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] poll for timed out futures in process group agent"


Follow up from https://github.com/pytorch/pytorch/pull/28392. Adds a background thread to `ProcessGroupAgent` that polls for timed out RPCs, and marks them as completed with a timeout exception if they have timed out. Also deletes the futures from the corresponding maps `futures_` and `futureTimeouts`. Unit tests are added to ensure that timed out RPCs are appropriately cleaned up.

The existing `futureTimeouts_` map is reworked so that the key is now the time (in ms) that the RPC would timeout, instead of when the RPC began. The reason for this is that it allows us to order futures which have different timeouts, which is a feature that we'd want to support in the future. A helper function `getRPCEndTime` can be used to retrieve the key into this map. 

Since we'd like to be able to quickly stop checking for timed out RPCs after looking at the first entry in the `futureTimeouts_` map, we cannot key by beginning time (an RPC with a much smaller timeout may later be created) or by the timeout itself (since an RPC with a large timeout would not be at the beginning, even if it is about to time out).

Also adds a `shutdown` variable to process group agent to control the shutting down of this background thread, which can eventually be extended to use for controlling a clean shutdown of process group agent.

Differential Revision: [D18434215](https://our.internmc.facebook.com/intern/diff/D18434215/)

[ghstack-poisoned]

rohan-varma commented:
Windows tests is broken, I think since this PR: https://github.com/pytorch/pytorch/pull/29747 - will work on a fix

pietern commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -70,13 +71,45 @@ class ProcessGroupAgent : public RpcAgent {
+          startTime_(startTime),
+          dstRank_(dstRank),
+          timeout_(timeout) {}
+    FutureInfo() {}
```
Beware: this allows for instances with garbage data. I'd either add member initializers here, or remove the default constructor entirely (if you're worried about inserting them into maps, use `std::move` or `emplace`).

rohan-varma replied:
Fixed in https://github.com/pytorch/pytorch/pull/30197

xush6528 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -281,9 +293,22 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
+      }
+      auto futureInfo = FutureInfo(future, futureStartTime, to.id_, timeout);
+      futures_[requestId] = futureInfo;
+      auto rpcEndTime = getRPCEndTime(futureInfo);
```
#30342

rohan-varma replied:
Yes, this would be a good optimization.

xush6528 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +482,98 @@ void ProcessGroupAgent::listenLoop() {
+    }
+
+    if (sleepTime == INFINITE_TIMEOUT) {
+      futureTimeoutCV_.wait(lock);
```
Should we add a `continue;` here? 

Since when being waken up from `::send(..)` call, the end times should not has arrived yet. There is no point of scanning the 2 maps?

xush6528 replied:
Doing this in https://github.com/pytorch/pytorch/pull/30355.

xush6528 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +482,98 @@ void ProcessGroupAgent::listenLoop() {
+}
+
+const std::vector<ProcessGroupAgent::FutureInfo> ProcessGroupAgent::
+    processTimedOutFutures() {
```
Notice this function is mutating the 2 maps that need the mutex's protection.

I guess it would be better if this function takes a lock, which is moved from the same lock used by `futureTimeoutCV_.wait_for(lock, sleepTime);`

And at the end of the function,
```cpp
lock.unlock()
futureCV_.notify_all();
```

xush6528 replied:
Doing this in https://github.com/pytorch/pytorch/pull/30355.

xush6528 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -449,6 +482,98 @@ void ProcessGroupAgent::listenLoop() {
+    // Do not hold the lock while marking futures completed, as markCompleted()
+    // could invoke callbacks.
+    lock.unlock();
+    for (const auto& timedOutFuture : timedOutFutures) {
```
Doing this in https://github.com/pytorch/pytorch/pull/30355.

rohan-varma replied:
I believe we don't hold the lock when notifyiing? It is unlocked on line 516.
